test/powershell/engine/ParameterBinding/ParameterBinding.Tests.ps1

```diff
@@ -424,3 +424,331 @@
+                [string]$Command,
+
+                [Parameter(Mandatory, ParameterSetName = "Command")]
+                [string]$ParameterName,
+
+                [Parameter(Mandatory, ParameterSetName = "Command")]
+                [object]$Argument
+            )
+            $ps.Commands.Clear()
+            $ps.Streams.ClearStreams()
+
+            if ($PSCmdlet.ParameterSetName -eq "Script") {
+                $ps.AddScript($Script).Invoke()
+            } else {
+                $ps.AddCommand($Command).AddParameter($ParameterName, $Argument).Invoke()
+            }
+        }
+
+        ## Helper command strings
+        $changeToConstrainedLanguage = '$ExecutionContext.SessionState.LanguageMode = "ConstrainedLanguage"'
+        $getLanguageMode = '$ExecutionContext.SessionState.LanguageMode'
+        $importScriptModule = "Import-Module $TestDrive\module.psm1"
+        $importCSharpModule = "Import-Module $asmFile"
+    }
+
+    AfterAll {
+        ## Set the LanguageMode to force rebuilding the type conversion cache.
+        ## This is needed because type conversions happen in the new powershell runspace with 'ConstrainedLanguage' mode
+        ## will be put in the type conversion cache, and that may affect the default session.
+        $ExecutionContext.SessionState.LanguageMode = "FullLanguage"
+    }
+
+    It "Custom type conversion in parameter binding is allowed in FullLanguage" {
+        ## Create a powershell instance for the test
+        $ps = [powershell]::Create()
+        try {
+            ## Import the modules in FullLanguage mode
+            Execute-Script -ps $ps -Script $importScriptModule
+            Execute-Script -ps $ps -Script $importCSharpModule
+
+            $languageMode = Execute-Script -ps $ps -Script $getLanguageMode
+            $languageMode | Should Be 'FullLanguage'
+
+            $result1 = Execute-Script -ps $ps -Script "Test-ScriptCmdlet -File fileToUse"
+            $result1 | Should Be "fileToUse"
+
+            $result2 = Execute-Script -ps $ps -Script "Test-ScriptFunction -File fileToUse"
+            $result2 | Should Be "fileToUse"
+
+            $result3 = Execute-Script -ps $ps -Script "Test-BinaryCmdlet -File fileToUse"
+            $result3 | Should Be "fileToUse"
+
+            ## If the conversion involves setting properties of an instance of the target type,
+            ## then it's disallowed even for trusted cmdlets
+            $hashValue = @{ FileName = "filename"; Arguments = "args" }
+            $psobjValue = [PSCustomObject] $hashValue
+
+            ## Test 'Test-ScriptCmdlet -StartInfo' with IDictionary and PSObject with properties
+            $result4 = Execute-Script -ps $ps -Command "Test-ScriptCmdlet" -ParameterName "StartInfo" -Argument $hashValue
+            $result4 | Should Be "filename"
+            $result5 = Execute-Script -ps $ps -Command "Test-ScriptCmdlet" -ParameterName "StartInfo" -Argument $psobjValue
```

Comment:
I am confused.  It looks like the conversion is performed since the result is StartInfo.FileName ("filename").  But the comment says this should be disallowed.

Reply:
Oh, I see we are in full language here.  Can you modify the comment?

Reply:
Good catch. I will fix the comments.